Jetpack Error UX: Resolve concerns about state sync

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes: https://github.com/Automattic/dotcom-forge/issues/4647

## Proposed Changes

This pr aims to decouple react query from our current state management.

cc: @flootr , @tyxla 

## Testing Instructions
1. Create a new WoA site.
2. Enable ssh.
3. Login to the site and add the following to your `wp-config`
```
if ( ! empty( $_GET['rest_route'] ) ) {
	http_response_code( 404 );
	die('not found');
}
```
4. Open developer console.
5. Visit eg `/posts` page
6. Ensure that you see an error.
7. Ensure that you see a request at `/test-connection`
8. Ensure that you see a request at `/jetpack-connection-health`
9. Reload the page
10. Ensure that `/jetpack-connection-health` endpoint is not called
11. Wait 5 minutes / or clear your IndexedDB cache
12. Ensure that `/jetpack-connection-health` endpoint is called again

Below is a video that shows the request caching in case of Jetpack health error:

https://github.com/Automattic/wp-calypso/assets/497103/9eb395fe-aac0-4d61-9241-38dfd2ab80a1



<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

kozer committed:
Jetpack Error UX: Resolve concerns about state sync

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health-request-error.js:
```diff
@@ -0,0 +1,19 @@
+		return null;
+	}
+
+	return state?.requestError;
```
Did you mean:

```
return siteState?.requestError;
```

if yes, then we could shorten this to:

```suggestion
	return state.jetpackConnectionHealth[ siteId ]?.requestError ?? null;
```

that way we will also always return `null` in the case of non-error, instead of a mix of `null` and `undefined`.

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,16 @@
+	}
+	const siteState = state.jetpackConnectionHealth[ siteId ];
+
+	return siteState?.connectionHealth;
```
Similar to above:

```suggestion
	return state.jetpackConnectionHealth[ siteId ]?.connectionHealth ?? null;
```

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health-request-error.js:
```diff
@@ -0,0 +1,19 @@
+ * Returns error code if the current site has possible Jetpack connection problem
+ * @param  {Object}  state         Global state tree
+ * @param  {number}  siteId        Site ID
+ * @returns {?string}              Connection error code
```
Did you mean `string` or `null`?

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,16 @@
+ * Returns true if the current site Jetpack connection was checked and is unhealthy
+ * @param  {Object}  state         Global state tree
+ * @param  {number}  siteId        Site ID
+ * @returns {?Object}             Whether the current site can have connection problem
```
Did you mean `Object|null`?

Also, the description appears to hint that it returns a boolean, but instead, it returns an object with health data.

tyxla commented on client/components/data/query-jetpack-connection-health/index.jsx:
```diff
@@ -0,0 +1,29 @@
+import isRequestingJetpackConnectionHealthStatus from 'calypso/state/jetpack-connection-health/selectors/is-requesting-jetpack-connection-health-status';
+
+const request = ( siteId ) => ( dispatch, getState ) => {
+	if ( ! isRequestingJetpackConnectionHealthStatus( getState(), siteId ) ) {
```
The `isRequestingJetpackConnectionHealthStatus` check seems obsolete. We'll trigger `request()` only once per site anyway because that's how `useEffect()` works - it will only retrigger if `siteId` changes. So the `isRequestingJetpackConnectionHealthStatus` will never be `true`, making that check unnecessary. This will make the query component way simpler. That also means we can remove the requesting status state subtree altogether, no?

kozer replied:
You are absolutely right about the check here. It is redundant. Updated. Thanks for your feedback!

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,23 +22,42 @@ import {
 	siteId: number;
 }
 
+const request = ( siteId ) => ( dispatch, getState ) => {
+	if ( ! isRequestingJetpackConnectionHealthStatus( getState(), siteId ) ) {
+		dispatch( requestJetpackConnectionHealthStatus( siteId ) );
+	}
+};
+
```
Why aren't we just using the query component here?

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,23 +22,42 @@ import {
+		getJetpackConnectionHealth( state, siteId )
+	);
+
+	useEffect( () => {
+		if ( ! siteId ) {
+			return;
+		}
+		dispatch( request( siteId ) );
+	}, [ dispatch, siteId ] );
```
We should totally use the query component here and reuse some code.

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,113 @@ import {
+ * @param {Object} action - action
+ * @returns {Object} updated state
+ */
+export const isLoading = withPersistence( ( state = false, action ) => {
```
Feels off to persist the `isLoading` state. Should be fine if we end up removing it as I suggested in another comment.

tyxla commented on client/state/action-types.ts:
```diff
@@ -373,6 +373,12 @@ export const JETPACK_CONNECTION_STATUS_REQUEST_FAILURE =
 	'JETPACK_CONNECTION_STATUS_REQUEST_SUCCESS';
+export const JETPACK_CONNECTION_HEALTH_REQUEST = 'JETPACK_CONNECTION_HEALTH_REQUEST';
+export const JETPACK_CONNECTION_HEALTH_REQUEST_RECEIVE = 'JETPACK_CONNECTION_HEALTH_RECEIVE';
+export const JETPACK_CONNECTION_HEALTH_REQUEST_SUCCESS =
```
Since this is only used for loading progress, and I suggested removing that, we should be able to ditch the `JETPACK_CONNECTION_HEALTH_REQUEST_SUCCESS` action type completely.

kozer replied:
I use that in the jetpack health banner as I mentioned below. Any ideas on how / if we can do better?

tyxla replied:
I left a comment above on how to get rid of its sole, and IMHO unnecessary usage.

kozer replied:
I see that now! I remove the unnecessary code! Thanks!

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,42 @@
+		},
+	},
+};
+export const jetpacConnectionHealth = {
```
There is a typo here:

```suggestion
export const jetpackConnectionHealth = {
```

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,42 @@
+export const schema = {
+	type: 'object',
+	patternProperties: {
+		'^\\d+$': jetpacConnectionHealth,
```
```suggestion
		'^\\d+$': jetpackConnectionHealth,
```

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,42 @@
+	type: [ 'string', 'null' ],
+};
+
+export const isLoading = {
```
Loading state shouldn't be persisted.

tyxla commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js:
```diff
@@ -14,13 +14,13 @@ export default function isJetpackConnectionProblem( state, siteId ) {
 	}
 
-	return connection_data.jetpack_connection_problem;
+	return siteState.connectionHealth.jetpack_connection_problem;
```
This one could be made shorter in similar ways as suggested above.

kozer replied:
I ended up not doing it like that as a possible `false` value is determined as falsy and so it ended up returning `null`

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Returns true if the current site Jetpack connection was checked and is unhealthy
+ * @param  {Object}  state         Global state tree
+ * @param  {number}  siteId        Site ID
```
If the site ID is always a number, why do we need the `siteId` check below? Is there a chance we'll have `siteId` of `0`?

tyxla commented on client/state/jetpack-connection-health/selectors/is-requesting-jetpack-connection-health-status.js:
```diff
@@ -0,0 +1,16 @@
+
+/**
+ * Returns true if we are currently making a request to retrieve the connection status. False otherwise.
+ * Returns null if the site is unknown, or there is no information yet.
```
This is currently incorrect; it will return `undefined` if there is no information about that site.

tyxla commented on client/state/jetpack-connection-health/selectors/is-requesting-jetpack-connection-health-status.js:
```diff
@@ -0,0 +1,16 @@
+ * Returns null if the site is unknown, or there is no information yet.
+ * @param  {Object}   state    Global state tree
+ * @param  {number}   siteId   The ID of the site we're querying
+ * @returns {?boolean}          Whether the connection status is being requested
```
Let's make sure this param reflects the truth and includes any additional field types.

tyxla commented on client/state/jetpack-connection-health/selectors/is-requesting-jetpack-connection-health-status.js:
```diff
@@ -0,0 +1,16 @@
+ * @returns {?boolean}          Whether the connection status is being requested
+ */
+export default function isRequestingJetpackConnectionHealthStatus( state, siteId ) {
+	if ( ! siteId ) {
```
Same as above for the potentially redundant `siteId` check

tyxla commented on client/state/action-types.ts:
```diff
@@ -373,6 +373,12 @@ export const JETPACK_CONNECTION_STATUS_REQUEST_FAILURE =
 export const JETPACK_CONNECTION_STATUS_REQUEST_SUCCESS =
 	'JETPACK_CONNECTION_STATUS_REQUEST_SUCCESS';
+export const JETPACK_CONNECTION_HEALTH_REQUEST = 'JETPACK_CONNECTION_HEALTH_REQUEST';
+export const JETPACK_CONNECTION_HEALTH_REQUEST_RECEIVE = 'JETPACK_CONNECTION_HEALTH_RECEIVE';
```
This action type seems unused.

Also it's a potentially difficult to track issue when the action type name isn't equal to the action type value.

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,113 @@ import {
+export const requestError = withPersistence( ( state = '', action ) => {
+	switch ( action.type ) {
+		case JETPACK_CONNECTION_HEALTH_REQUEST:
+			return '';
```
Should we use `null` instead of an empty string? I don't see us doing a separate check to check for an empty string.

kozer replied:
I think it's ok to use an empty string here, instead of a null. An empty string is a false value, and I prefer having default values of a type instead of null or undefined. I followed on every property of this reducer

tyxla replied:
I disagree on this; `null` and `undefined` are both falsy too, and being able to differentiate by type vs only by value whether we have an error is better IMHO, because we rarely handle empty strings specifically, while we handle `null` and `undefined` specifically.

tyxla commented on client/state/jetpack-connection-health/selectors/is-requesting-jetpack-connection-health-status.js:
```diff
@@ -0,0 +1,16 @@
+ * @param  {number}   siteId   The ID of the site we're querying
+ * @returns {?boolean}          Whether the connection status is being requested
+ */
+export default function isRequestingJetpackConnectionHealthStatus( state, siteId ) {
```
Seems like as I mentioned above we'll be able to remove this selector and its corresponding state subtree altogether.

kozer replied:
This is used at `JetpackConnectionHealthBanner` component.
Thanks for your comment!

tyxla replied:
I saw the usage, but we could still consider removing it. Often from the perspective of the consumer component, whether the data is being requested is similar to whether the data is missing - the component doesn't care and treats both cases the same way.

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,16 @@
+import 'calypso/state/jetpack-connection-health/init';
+
+/**
+ * Returns true if the current site Jetpack connection was checked and is unhealthy
```
This description doesn't seem to be accurate; the selector actually returns the jetpack connection health information instead.

tyxla reviewed (commented):
Thanks for working on it! 

I think we can simplify a great bit of it, see the separate comments. Also, there is a bit of imprecision in how we handle certain data types. 

Other than that, I think this is heading in the right direction!

kozer committed:
Update: Minor refactor based on comments

kozer commented:
@tyxla, first of all thank you very much for your detailed review. It was intended just to check if it's in the right direction or not, however, your comments helped me to find out the issues that were lying around! Thank you very much!
As you think this is the right direction. I'll fix tests etc and ping you again for a final review if it's ok. 
Thanks again!

tyxla commented:
> @tyxla, first of all thank you very much for your detailed review. It was intended just to check if it's in the right direction or not, however, your comments helped me to find out the issues that were lying around! Thank you very much! As you think this is the right direction. I'll fix tests etc and ping you again for a final review if it's ok. Thanks again!

Of course, let me know when it's ready for another review 👍

kozer committed:
Update: Minor refactor based on comments + add tests

kozer committed:
Update: avoid storing loading state

kozer committed:
Update: Remove isLoading from schema

kozer committed:
Fix: typo on export

kozer committed:
Fix: failing tests

kozer committed:
Update: Try to fix lint errors

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

kozer commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,23 +23,35 @@ import {
+	);
+
+	useEffect( () => {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
How we could remove this ignore?

tyxla replied:
Why do you need the ignore in the first place?

kozer replied:
It seems that there was a ts error in my IDE, which now is not there. Strange. We don't need this `ts-ignore`, so I'll remove it.

kozer commented:
@tyxla I did the proper changes and add/fix tests.
However, it seems that e2e tests are failing ( Desktop ). It doesn't seem to be this pr's fault, but even if I rerun them many times, I keep getting the same results.
Any idea why this is happening?
PS: I  also provided some comments for some recommendations you made. When you have some time, please take a look. Thanks.

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,23 +23,35 @@ import {
+	}, [ dispatch, siteId ] );
 
 	if (
 		isLoadingJetpackConnectionHealth ||
```
My previous comment on this is still valid: there's no good reason to use `isLoadingJetpackConnectionHealth` at all, we can remove it. We can just rely on `jetpackConnectionHealth?.error` below - if it's nullish, then we don't need to do anything in this component anyway. That will allow us to remove:

* The `isLoadingJetpackConnectionHealth()` selector
* The `isLoading` reducer subtree
* The `JETPACK_CONNECTION_HEALTH_REQUEST_SUCCESS` action
* The related tests

kozer replied:
I see your point now that I am thinking about this better. Thanks for your feedback! I'll do the appropriate changes.

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,112 @@ import {
+export const requestError = withPersistence( ( state = '', action ) => {
+	switch ( action.type ) {
+		case JETPACK_CONNECTION_HEALTH_REQUEST:
+			return null;
```
I still don't quite understand why we maintain an empty string as a default state, and a `null` as a potential value. Why do we need to have 2 different "empty" states? I actually commented about this already in my previous review.

tyxla reviewed (commented):
Thanks for your work here so far!

I can see a bunch of the comments I left in my previous review were unaddressed - 
I think if you go through them one by one and address them, or at least reply about why you haven't addressed them, it will be easier to move forward. Otherwise, we might be going in circles since I might be repeating myself. 

In a nutshell, there are some issues with the code that I still think we can still address:

* Some unnecessary complication that can be avoided (the loading state)
* Data type incompatibilities - multiple empty states or incompatible empty string vs null vs undefined types. 

I made sure to comment again so there's no extra confusion, but please let me know if you have any additional questions. 

Otherwise, I think once we clean those up, this should be pretty close!

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

kozer committed:
Update: based on review comments

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

kozer committed:
Update: Fix lint errors

kozer commented:
@tyxla I updated the PR.
What I did was:
- Remove unnecessary `isLoading` state.
- Change all state values to be their falsy equivalent ( empty string instead of null, empty object instead of null )
- Update the docs of the selectors
- Fix tests

tyxla commented on client/components/jetpack/connection-health/test/index.jsx:
```diff
@@ -48,9 +37,18 @@ const render = ( el, options = {} ) =>
 describe( 'JetpackConnectionHealthBanner', () => {
 	describe( 'component rendering', () => {
 		test( 'shows generic message', () => {
+			jest.useFakeTimers().setSystemTime( Date.now() );
```
Can this be done without fake timers? Seems like there is no good reason to use them since we're not really testing timing and precision. Or if there is a good reason to use fake timers, can you elaborate?

kozer replied:
I wanted to ensure that this test will always be fixed to a date and the check we do if 5 minutes would pass will always succeed. Now I'm thinking more about it, there is no way that between setting the `lastRequestTime` and the check that 5 minutes will pass. I removed the fake timers.
Thank you for your comment!

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,33 +22,35 @@ import {
-		isErrorCheckJetpackConnectionHealth ||
-		jetpackConnectionHealth?.is_healthy
-	) {
+	const isErrorCheckJetpackConnectionHealth = useSelector( ( state ) =>
```
We may need to rename this constant. At first glance, there is confusion between request error and connection health error, and we might need to be extra precise about naming to reduce that confusion.

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,11 @@
+ * @returns {{ is_healthy?: boolean, error?: string, jetpack_connection_problem?: boolean }}          Jetpack connection health
+ */
+export default function getJetpackConnectionHealth( state, siteId ) {
+	return state.jetpackConnectionHealth[ siteId ]?.connectionHealth ?? {};
```
Creating a new object every time may result in unnecessary rerenders.

kozer replied:
You are right on this one. I'll make that `null` again. Is it ok?

tyxla replied:
Sounds good 👍 FWIW empty objects, empty strings, empty arrays etc. need additional evaluation to ensure they're actually empty, while `null` and `undefined` make that immediately clear.

kozer replied:
The only place we did an evaluation was inside `JetpackConnectionHealthBanner` where we didn't check for the object itself but for its `is_healthy` property, so we would be ok. However, I see the point that if we reuse that in the future, one has to dig into the selector to see why a simple check would not work.

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,11 @@
+ * @returns {{ is_healthy?: boolean, error?: string, jetpack_connection_problem?: boolean }}          Jetpack connection health
+ */
+export default function getJetpackConnectionHealth( state, siteId ) {
+	return state.jetpackConnectionHealth[ siteId ]?.connectionHealth ?? {};
```
Given that the sole usage of this selector we access with `?.`, does it make sense to return `null` or `undefined` instead of an empty object? Empty objects will often not represent "empty state" in simple conditions, since an empty object is evaluated as `true` (`Boolean( {} ) === true`)

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +31,56 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+	const currentState = getState();
+	const lastRequestTime = currentState.jetpackConnectionHealth[ siteId ]?.lastRequestTime;
+	if ( lastRequestTime && Date.now() - lastRequestTime < 1000 * 60 * 5 ) {
+		return currentState?.jetpackConnectionHealth[ siteId ]?.connectionHealth?.is_healthy;
```
Why does this thunk return some value? What are we expecting to do with it?

kozer replied:
As it turned out, we didn't do anything with this one. I'll simply return instead of return the current value.

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +31,56 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+export const requestJetpackConnectionHealthStatus = ( siteId ) => ( dispatch, getState ) => {
+	const currentState = getState();
+	const lastRequestTime = currentState.jetpackConnectionHealth[ siteId ]?.lastRequestTime;
+	if ( lastRequestTime && Date.now() - lastRequestTime < 1000 * 60 * 5 ) {
```
The stale time can be moved to a separate setting in a constant above in this module.

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,95 @@ import {
-					error: '',
-				},
+				jetpack_connection_problem: false,
+				is_healthy: true,
```
Why do we need the extra `is_healthy` flag? Can we just rely on the `error` field? If it's empty (null, undefined, or empty string, whatever), then the connection is considered healthy? And the other way around - if there is an error, then it's not healthy?

kozer replied:
Yes you are right! The code is much simpler now! Thanks!

tyxla commented on client/components/jetpack/connection-health/test/index.jsx:
```diff
@@ -207,11 +277,20 @@ describe( 'JetpackConnectionHealthBanner', () => {
+							error,
+							jetpack_connection_problem: true,
+						},
+					},
```
If this is the same all along for most (or all) tests, can we move it outside to a fixture constant and reuse it?

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +31,56 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+				dispatch( setJetpackConnectionHealthy( siteId ) );
+			}
+			if ( ! isHealthy && ! reduxIsUnhealthy ) {
+				dispatch( setJetpackConnectionUnhealthy( siteId, error ?? '' ) );
```
In which instance will we have an unhealthy connection but with no error?

kozer replied:
I checked backend code and I saw that there is no such case. I removed the check.

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +31,56 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+		} )
+		.catch( ( error ) => {
+			dispatch( {
+				type: JETPACK_CONNECTION_HEALTH_REQUEST_FAILURE,
```
What types of errors does this handle? Is it just regular request errors? Can we just do nothing if we have a request failure, and as a result, clean this part of the state up?

kozer replied:
In case where jetpack connection endpoint returns a healthy response, and the health status request fails, we will keep showing the error message to the user instead of not showing him anything. This is why we check that inside `JetpackConnectionHealthBanner` component.

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +31,56 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+ * @param {number} siteId The site id to which the status belongs
+ * @returns {Object} An action object
+ */
+export const setJetpackConnectionMaybeUnhealthy = ( siteId ) => ( {
```
I find some additional confusion here - we have a few states:

- Healthy connection
- Unhealthy connection
- Maybe unhealthy connection
- Connection request error

Can we document them so someone not familiar with those concepts will have a better understanding?

kozer replied:
I provided some extra information about those actions ( especially for the maybe unhealthy which I find the most confusing ).
Please comment here if you want some extra information.

tyxla commented on client/state/jetpack-connection-health/selectors/test/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,26 @@
+			jetpack_connection_problem: true,
+		} );
+	} );
+	test( 'should return null if no connection health object exists', () => {
```
This returns an empty object actually.

kozer replied:
I changed it to return `null` to avoid unnecessary rerenders, as discussed.

tyxla commented on client/state/jetpack-connection-health/selectors/test/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,26 @@
+			jetpack_connection_problem: true,
+		} );
+	} );
+	test( 'should return null if no connection health object exists', () => {
```
Should we also have a test that covers when we don't have connection health data for the site at all in the state tree?

tyxla commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js:
```diff
@@ -8,19 +8,19 @@ import 'calypso/state/jetpack-connection-health/init';
 	}
 
-	return connection_data.jetpack_connection_problem;
+	return siteState.connectionHealth.jetpack_connection_problem;
```
Is there a possibility this will return `undefined` because `connectionHealth` is an empty object and doesn't contain a `jetpack_connection_problem` property?

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +31,56 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+			if ( isHealthy && reduxIsUnhealthy ) {
+				dispatch( setJetpackConnectionHealthy( siteId ) );
+			}
+			if ( ! isHealthy && ! reduxIsUnhealthy ) {
```
The `isJetpackConnectionUnhealthy()` selector can return `null` and `false` and we don't seem to make a difference with this `! reduxIsUnhealthy` check. Is that something to worry about?

kozer replied:
I checked it and the return of null is not necessary. Returned false instead. Update the code to reflect that.

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,37 @@
+	type: 'number',
+};
+
+export const connectionHealth = {
```
It appears that we only consume the `schema` directly and the other schema pieces don't need to be exported from this module.

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,37 @@
+};
+
+export const lastRequestTime = {
+	type: 'number',
```
`lastRequestTime` can be `null` by default, should we cover that type too?

tyxla commented on client/state/jetpack-connection-health/selectors/test/get-jetpack-connection-health-request-error.js:
```diff
@@ -0,0 +1,24 @@
+		const output = getJetpackConnectionHealthRequestError( stateIn, siteId );
+		expect( output ).toBe( 'test' );
+	} );
+	test( 'should return null if no error exists', () => {
```
Do we need a test also if that site ID doesn't exist in the state at all?

tyxla commented on client/state/jetpack-connection-health/selectors/test/get-jetpack-connection-health-request-error.js:
```diff
@@ -0,0 +1,24 @@
+		};
+		const siteId = 123456;
+		const output = getJetpackConnectionHealthRequestError( stateIn, siteId );
+		expect( output ).toBe( '' );
```
The test promises to return `null` but returns an empty string?

tyxla commented on client/state/jetpack-connection-health/test/actions.js:
```diff
@@ -43,4 +51,91 @@ describe( 'action', () => {
+		} );
+
+		test( 'should set connection health to true if the connection api has failed but the health api finds the connection health ok.', async () => {
+			const siteId = 1;
```
Let's not use `siteId` of `1` please. I recommend going with some random number

tyxla commented on client/state/jetpack-connection-health/selectors/test/is-jetpack-connection-problem.js:
```diff
@@ -25,7 +25,7 @@ describe( 'isJetpackConnectionProblem()', () => {
 
 	test( 'should return false if the site Jetpack connection problem is unknown', () => {
```
Some of those tests are also wrong - the test claims to return `false` but returns `null`.

kozer replied:
Fixed!

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,37 @@
+};
+export const jetpackConnectionHealth = {
+	type: 'object',
+	properties: {
```
Should we mark any of those properties as required?

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,37 @@
+};
+export const jetpackConnectionHealth = {
+	type: 'object',
+	properties: {
```
Currently, extra properties will be allowed to be persisted. Should we add `additionalProperties: false` to disallow that?

kozer replied:
TIL. Thanks about that!

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,95 @@ import {
+import {
+	combineReducers,
+	withSchemaValidation,
+	keyedReducer,
```
Nit: we usually sort imports alphabetically.

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,95 @@ import {
+} );
+
+/**
+ * `Reducer` function which handles request/response actions concerning Jetpack connection requests
```
Not really request/response but rather request errors?

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,95 @@ import {
+ * `Reducer` function which handles tracking the last request time for Jetpack connection health
+ * @param {Object|null} state - current state
+ * @param {Object} action - action
+ * @returns {Object|null} updated state
```
Should we be more specific about the type of object?

tyxla reviewed (commented):
Thanks for your continuous hard work on this! 

I think we're moving in the right direction, but there are a few details to iron out. See comments below.

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

kozer committed:
Update: Remove unecessary fake timers

kozer committed:
Update: Remove unecessary fake timers from plugins-browser tests

kozer committed:
Update: jsdocs and selector default return value

kozer commented:
@tyxla I updated based on the latest comments. I also added a video showing the request caching.

kozer committed:
Update: Fix failing tests

kozer committed:
Update: Fix failing tests

kozer commented:
@tyxla I updated based on your latest comments. I also added a video showing the request caching.
More specifically:
- Updated jsdocs and added `additionalProperties` on reducer schema. Change import order in reducer.
- Renamed `isErrorCheckJetpackConnectionHealth` to `isJetpackConnectionHealthAPIError`.
- Changed connection health selector to return `null` instead of an empty object.
- Removed fake timers from tests.
- Change thunk to just return instead of returning the current value and updated failing test.

tyxla commented:
@kozer thanks. There are a bunch of other comments that seem unaddressed though. I assume you're still working on them?

kozer committed:
Update: move stale time to a constant

kozer committed:
Update: remove unecessary return of null. Return a boolean instead

kozer committed:
Update: remove is_healthy and use error code instead

kozer committed:
Update: schema for lastRequestTime

kozer committed:
Update: remove unecessary exports from schema

kozer committed:
Update: simplify test structure

kozer committed:
Update: Minor refactor and fix tests

kozer committed:
Update: Add some extra explanation at the Jetpack connection health actions

kozer committed:
Update: shema and fix a bug in reducer

kozer commented:
> There are a bunch of other comments that seem unaddressed though. I assume you're still working on them?
Yes, I was. I updated all the code based on your comments. I also tested it and it seems it works.
I'll be AFK till the start of the year, so no hurry on this one. Thank you very much for all the effort you put on commenting on this one @tyxla .

tyxla commented:
Thank you for the hard work on it @kozer! I'm planning to do another round of reviews, but it may as well be in 2024 since I'll also be AFK.

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,33 +22,35 @@ import {
+	);
+
+	useEffect( () => {
+		dispatch( requestJetpackConnectionHealthStatus( siteId ) );
```
Note on the way the freshness works when using schema. The data will be checked upon during hydration (during page load) and thrown away by our persistence mechanism, if it's stale. However, if the user stays on the page for hours, the health status will never be refreshed. Is this how you expect/want this to work?

kozer replied:
If the user moves on different pages, the health request will kick in eventually. 
If not, it's ok to be as is, as this was the functionality it was implemented before.

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,93 @@ import {
+);
 
-export default withStorageKey( 'jetpackConnectionHealth', jetpackConnectionHealth );
+export default withStorageKey( 'jetpackConnectionHealth', withSchemaValidation( schema, reducer ) );
```
The reducer structural changes will need some related unit test adjustments, particularly for the Jetpack Connect tests.

kozer replied:
@tyxla you mean the tests that were failing right? They are now fixed!

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,33 +22,35 @@ import {
 	}
 
-	const errorType = jetpackConnectionHealth?.error ?? '';
+	const errorType = jetpackConnectionHealth?.error || '';
```
Why is that change necessary? What difference does it make?

kozer replied:
No difference. I guess my vim changed that for some reason. Thanks for catching this one!

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +37,66 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+export const requestJetpackConnectionHealthStatus = ( siteId ) => ( dispatch, getState ) => {
+	const currentState = getState();
+	const lastRequestTime = currentState.jetpackConnectionHealth[ siteId ]?.lastRequestTime;
+	if ( lastRequestTime && Date.now() - lastRequestTime < STALE_CONNECTION_HEALTH_THRESHOLD ) {
```
I find it unexpected that I call a `requestJetpackConnectionHealthStatus()` action, but it won't request the connection status if it has recently been requested (AKA throttling). So, should this limitation be abstracted somewhere else, outside the data layer, allowing us to request fresh health status if we want to?

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +37,66 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+		} )
+		.catch( ( error ) => {
+			dispatch( {
+				type: JETPACK_CONNECTION_HEALTH_REQUEST_FAILURE,
```
Should this be moved to an action creator for consistency?

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -28,6 +26,8 @@ export const setJetpackConnectionHealthy = ( siteId ) => ( {
 /**
  * Sets the Jetpack connection status to unhealthy along with error code.
+ *
+ * This action is called when the Jetpack health status API returns an jetpack error.
```
Minor grammar nit.

```suggestion
 * This action is called when the Jetpack health status API returns a Jetpack error.
```

Also, should we be a bit more explicit about what a "Jetpack error" is in this case?

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +37,66 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
 } );
+
+/**
+ * Requests the Jetpack connection status from the server
```
```suggestion
 * Requests the Jetpack connection health status from the server
```

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +37,66 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+/**
+ * Requests the Jetpack connection status from the server
+ *
+ * This is called when the Jetpack connection is maybe unhealthy and we want to confirm
```
Aren't we calling it regardless of whether it's healthy or not? We don't know if it's healthy, that's why we're calling it, right?

kozer replied:
No, this is not the case, and I guess it's a bit confusing. There is a jetpack connection endpoint that is getting an error originally either through `fetchModuleList` or `failedJITM`.
Only then we call that API to get a better understanding of the error.
The reason we did it like this is that this API call is expensive, so we wanted to minimize when it's called and when it's not.

tyxla replied:
Thanks for elaborating. I guess this is a good opportunity to expand that comment with what you just explained.

kozer replied:
It's already explained in `setJetpackConnectionMaybeUnhealthy`, but I will add it here as well for better understanding.

tyxla commented on client/state/jetpack-connection-health/actions.js:
```diff
@@ -37,3 +37,66 @@ export const setJetpackConnectionUnhealthy = ( siteId, errorCode ) => ( {
+ */
+export const requestJetpackConnectionHealthStatus = ( siteId ) => ( dispatch, getState ) => {
+	const currentState = getState();
+	const lastRequestTime = currentState.jetpackConnectionHealth[ siteId ]?.lastRequestTime;
```
This should likely be abstracted to its own selector.

tyxla commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -17,33 +22,35 @@ import {
+		dispatch( requestJetpackConnectionHealthStatus( siteId ) );
+	}, [ dispatch, siteId ] );
+
+	if ( isJetpackConnectionHealthAPIError || jetpackConnectionHealth?.is_healthy ) {
```
Did you mean to use `! jetpackConnectionHealth?.jetpack_connection_problem` instead of `jetpackConnectionHealth?.is_healthy` here? I thought we're getting rid of `is_healthy`?

kozer replied:
I was meant to use the error here. Thanks for catching this one!

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,93 @@ import {
+ * `Reducer` function which handles Jetpack connection health state changes
+ * @param {Object} state - current state
+ * @param {Object} action - action
+ * @returns {{ is_healthy?: boolean, error?: string, jetpack_connection_problem?: boolean }} updated state - Jetpack connection health
```
Are we getting rid of `is_healthy`?

tyxla commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,93 @@ import {
+export const requestError = withPersistence( ( state = '', action ) => {
+	switch ( action.type ) {
+		case JETPACK_CONNECTION_HEALTH_REQUEST:
+			return '';
```
I still don't understand why an empty string is better than `null` or `undefined` here.

kozer replied:
I switched to null. The reason was that at some point I had encountered an issue, where when the state was `null` it wasn't saved in the localStorage.
I tested it though and it seems that the functionality works with `null` so it didn't make sense to fallback to an empty string instead.

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,39 @@
+const requestError = {
+	type: [ 'string', 'null' ],
```
This is currently only `string` right? I thought you intentionally stopped using `null` but I'm confused now.

kozer replied:
Switched to null, as discussed above! Sorry for the confusion!

tyxla commented on client/state/jetpack-connection-health/schema.js:
```diff
@@ -0,0 +1,39 @@
+const connectionHealth = {
+	type: 'object',
+	properties: {
+		is_healthy: {
```
Is this one still necessary?

kozer replied:
It was a leftover! Removed! Thanks!

tyxla commented on client/state/jetpack-connection-health/selectors/get-jetpack-connection-health.js:
```diff
@@ -0,0 +1,11 @@
+ * Returns Jetpack connection health for a site.
+ * @param  {Object}  state         Global state tree
+ * @param  {number}  siteId        Site ID
+ * @returns {{ is_healthy?: boolean, error?: string, jetpack_connection_problem?: boolean } | null}          Jetpack connection health
```
Are we still using `is_healthy`?

kozer replied:
It was a leftover! Fixed now! Thanks!

tyxla commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js:
```diff
@@ -8,19 +8,16 @@ import 'calypso/state/jetpack-connection-health/init';
-		return null;
-	}
-	const connection_data = state.jetpackConnectionHealth?.[ siteId ];
+	const siteState = state.jetpackConnectionHealth[ siteId ];
```
Should we use the `getJetpackConnectionHealth()` selector here?

tyxla commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-unhealthy.js:
```diff
@@ -4,19 +4,19 @@ import 'calypso/state/jetpack-connection-health/init';
 export default function isJetpackConnectionUnhealthy( state, siteId ) {
-	if ( ! siteId ) {
-		return null;
+	const siteState = state.jetpackConnectionHealth[ siteId ];
```
Should we use `getJetpackConnectionHealth()`?

tyxla commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-unhealthy.js:
```diff
@@ -4,19 +4,19 @@ import 'calypso/state/jetpack-connection-health/init';
-	return (
-		true === connection_data.jetpack_connection_problem && false === connection_data.is_healthy
-	);
+	return true === siteState.connectionHealth.jetpack_connection_problem && '' !== error;
```
This empty string check is a bit messed up. Would have been much better if we deal with `null` or `undefined` in that case.

tyxla commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-unhealthy.js:
```diff
@@ -4,19 +4,19 @@ import 'calypso/state/jetpack-connection-health/init';
-		return null;
+	let error = siteState.connectionHealth.error;
+	if ( siteState.connectionHealth.error === undefined ) {
+		error = '';
```
Why is this necessary? Seems like a code smell.

kozer replied:
Refactored the code and removed the code smell.

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

kozer committed:
Update: minor changes in docs and change jetpack_connection_problem if fetchModules or jitm succeeds

kozer committed:
Update: Move lastrequest time to a selector

kozer committed:
Update: move throttling to a different selector

kozer committed:
Update: better docs for action creators

kozer committed:
Update: Use selectore instead of direct usage of state

kozer committed:
Update: Use connection health selector instead of direct usage of state

kozer committed:
Update: Move request failure to it's own action creator and update tests

kozer committed:
Docs: health status reducer README

kozer committed:
Update: fix failing tests

kozer committed:
Update: fix failing tests

kozer committed:
Update: fix test case

kozer committed:
Update: Cover edge cases in tests

kozer commented:
@tyxla thank you for your comments and all your efforts in this review. I updated the code with your latest request changes, fixed the tests, and tested it in a real scenario.
When you have the time, please give it another look. Thank you very much!

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

tyxla commented:
> When you have the time, please give it another look. Thank you very much!
Thank you! Certainly, will try to set some time aside for another review tomorrow.

tyxla reviewed (commented):
Nothing further on code from my perspective. Thanks for all the hard work, @kozer 🙌 

Could you get someone from your team to do a final test and approval before shipping? Thank you.

kozer commented:
> Nothing further on code from my perspective. Thanks for all the hard work, @kozer 🙌
> 
> Could you get someone from your team to do a final test and approval before shipping? Thank you.

Thank you as well for the collaboration, and more importantly for the back-and-forths we had!

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq

sejas commented on client/components/jetpack/connection-health/test/index.jsx:
```diff
@@ -61,14 +63,8 @@ describe( 'JetpackConnectionHealthBanner', () => {
-				},
-			};
-
+			const error = DATABASE_ERROR;
+			const initialState = getJetpackConnectionHealthFixture( 1, error );
```
Nice simplification for the initialState.

sejas commented on client/state/jetpack-connection-health/reducer.js:
```diff
@@ -3,52 +3,93 @@ import {
+	JETPACK_CONNECTION_HEALTH_REQUEST,
+	JETPACK_CONNECTION_HEALTH_REQUEST_FAILURE,
 } from 'calypso/state/action-types';
+import {
+	combineReducers,
+	keyedReducer,
+	withPersistence,
+	withSchemaValidation,
+} from 'calypso/state/utils';
+import { schema } from './schema';
 
-export const jetpackConnectionHealth = ( state = {}, action ) => {
+/**
+ * `Reducer` function which handles Jetpack connection health state changes
+ * @param {Object} state - current state
+ * @param {Object} action - action
+ * @returns {{ error?: string, jetpack_connection_problem?: boolean }} updated state - Jetpack connection health
+ */
+export const connectionHealth = withPersistence( ( state = {}, action ) => {
```
Out of curiosity, why are we renaming this function?

kozer replied:
As this file is under the folder `jetpack-connection-health` and this reducer is part of  `jetpackConnectionHealth` object, I found it a bit redundant to name it `jetpackConnectionHealth`.

sejas commented on client/components/jetpack/connection-health/test/index.jsx:
```diff
@@ -3,7 +3,7 @@
 import { screen } from '@testing-library/react';
 import { merge } from 'lodash';
-import { jetpackConnectionHealth } from 'calypso/state/jetpack-connection-health/reducer';
+import { reducer as jetpackConnectionHealth } from 'calypso/state/jetpack-connection-health/reducer';
```
Other alternatives I tried and are also working:
```
import { connectionHealth as jetpackConnectionHealth } from 'calypso/state/jetpack-connection-health/reducer';
```
```
import jetpackConnectionHealth from 'calypso/state/jetpack-connection-health/reducer';
```

kozer replied:
I wanted the full reducer here, without the `storage` part, and this is why I imported it like that, so this excludes the second option.
Regarding the first one, this would import only the `connectionHealth` reducer part, which for now is fine ( as I didn't add more tests regarding the other parts of the reducer ), but as those tests supposed to test the whole reducer object, I'll leave it as is! Thanks for pointing this out!

sejas reviewed (approved):
I created an atomic site and broke the connection with your snippet in wp-config.php.
I confirm I see the errors in `/posts/$siteSlug` and other pages like `/home/$siteSlug`.
I confirm the two requests are made and `/v2/sites/227982990/jetpack-connection-health` is only repeated after 4 minutes.

Great job! I added some comments about `jetpackConnectionHealth` export.

<img width="1728" alt="Screenshot 2024-01-12 at 19 37 33" src="https://github.com/Automattic/wp-calypso/assets/779993/d311a680-62dd-410f-b832-bfc0aad8be88">
<img width="1728" alt="Screenshot 2024-01-12 at 19 33 06" src="https://github.com/Automattic/wp-calypso/assets/779993/d7cbdde7-0ceb-422c-a92f-79de1f0370b2">

kozer committed:
Merge branch 'trunk' into fix/state_sync_jetpack_rq
